x-pack/plugins/observability/server/services/slo/create_slo.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import uuid from 'uuid';
+
+import { SLO } from '../../types/models';
+import { ResourceInstaller } from './resource_installer';
+import { SLORepository } from './slo_repository';
+import { TransformInstaller } from './transform_installer';
+
+import { CreateSLOParams, CreateSLOResponse } from '../../types/schema';
+
+export class CreateSLO {
+  constructor(
+    private resourceInstaller: ResourceInstaller,
+    private repository: SLORepository,
+    private transformInstaller: TransformInstaller,
+    private spaceId: string
+  ) {}
+
+  public async execute(sloParams: CreateSLOParams): Promise<CreateSLOResponse> {
+    const slo = this.toSLO(sloParams);
+
+    await this.resourceInstaller.ensureCommonResourcesInstalled(this.spaceId);
+    await this.repository.save(slo);
+
+    try {
+      await this.transformInstaller.installAndStartTransform(slo, this.spaceId);
+    } catch (err) {
+      await this.repository.deleteById(slo.id);
+      throw err;
+    }
+
+    return this.toResponse(slo);
+  }
+
+  private toSLO(sloParams: CreateSLOParams): SLO {
+    return {
+      ...sloParams,
+      id: uuid.v1(),
+      settings: {
+        destination_index: sloParams.settings?.destination_index,
+      },
+    };
+  }
+
+  private toResponse(slo: SLO): CreateSLOResponse {
```

Comment:
nit: I believe this function is needed to return only the id of a created SLO  as responded for `POST /api/observability/slos`
However, `CreateSLO` is a service, AFIAK,  services should be oblivious about who would call them. Having the `toResponse ` as a member class makes `CreateSLO` aware of the route implementation.

Reply:
TLDR: I think `toResponse` is misleading and instead should be read as `toDTO`.

It's not obvious because all services are within the same folder but `CreateSLO` is an application service [(as opposed to a domain service)](https://khalilstemmler.com/articles/software-design-architecture/domain-driven-design-vs-clean-architecture/)

It is supposed to be used at the boundary of the domain to fulfill a specific use case, e.g. from a REST handler, Queue consumer, etc...
It transforms the infrastructure specific data model (http body, queue message, ...) into the internal domain model, and orchestrates the different domain services to fulfill the use case, then transforms the domain model to a DTO model.
An application service avoids leaking the domain model to the outside world.

`toResponse` is not specific to a route implementation (there is nothing related to status code, or other implementation details of a route handler). It only produces a DTO to avoid leaking the domain model to the infrastructure layer (HTTP router in this case).
That being said, I see that `toResponse` can be misleading and interpreted as "an HTTP response". But what it does is really transforming the domain model to a DTO that is not specific to any implementation details. 

👉🏻  Maybe a less misleading name could be `toDTO`?

Happy to continue to chat about this online